include/envoy/upstream/load_balancer.h

```diff
@@ -43,6 +43,27 @@ class LoadBalancerContext {
    * balancing.
    */
   virtual const Http::HeaderMap* downstreamHeaders() const PURE;
+
+  /**
+   * Called to retrieve an optional pre priority selection host predicate. If present, this will be
+   * used to rebuild the per priority load data which only includes hosts that match the predicate.
+   *
+   */
+  virtual absl::optional<std::function<bool(uint32_t, const Host&)>>
+  prePrioritySelectionFilter() PURE;
+
+  /**
+   * Called to determine whether we should reperform host selection. The load balancer
+   * will retry host selection until either the postHostSelectionFilter returns true
+   * or hostSelectionRetryCount is reached.
+   */
+  virtual bool postHostSelectionFilter(const Host& host) PURE;
+
+  /**
+   * Called to determine how many times host selection should be retried until the filter is
```

Comment:
I'm wondering if this is the right API to begin with. In which situation do you want to use `postHostSelectionFilter/hostSElectionRetryCount` vs. `prePrioritySelectionFilter`? Seems the former should be expressive enough to capture the latter. The disadvantage of the latter is that you could ultimately be retrying arbitrary numbers of time and/or not have the filter applied effectively.

Reply:
The original motivation for having them be separate was to allow mutating the `per_priority_load` separately from rejecting individual hosts in order to make priorities that would otherwise see no traffic receive traffic (e.g. P0 is completely healthy, we want to route to P1). 

That said, I can't actually think of a case where you don't also want to adjust the per_priority_load to exclude the hosts rejected by `prePrioritySelectionFilter`, so I think it makes sense to combine these into a single predicate. The only case I can think of is when you only have one priority, but I think we could handle that instead of delegating that to the callback.

Perhaps it makes more sense to do

```
bool shouldSelectAnotherPriority(uint32_t priority);
bool shouldSelectAnotherHost(const Host& host);
```

That way the host predicate can be shared between the pre-priority and post-host selection. The separate priority predicate means that we can exclude an entire `HostSet` without iteration over it to filter it, which seems like a nice little perf gain.

Reply:
Thinking about it some more - I can kinda imagine there being a use case where one might not want to pay the cost of having to rebuild the priority table (e.g. if they have a very large number of hosts), but still make a best effort attempt at avoiding routing to some hosts. 

Maybe the predicate should be the same, but a separate flag determines whether to apply it prior to priority selection?

Reply:
@snowp yes perf is basically the reason that I suggested the pre/post filter options. IMO, the way to implement "don't retry to same host twice" is to implement a trivial comparison filter with some number of max selection attempts. This is basically an O(1) operation which doesn't require any allocations. I agree that it might be better to just offer up an interface of some type that allows the implementer to decide what they want to do? For example the could pass through the existing priority sets or return a reference to a new one that is internally stored?

Reply:
Actually, given that the priority selection is based on `per_priority_load` which is maintained separately from the priority set, it should probably be something more like: 

```
const std::vector<uint32_t>& recalculatePerPriorityLoad(const PrioritySet& priority_set, const std::vector<unit32_t>& original_priority_load);
```

Reply:
@snowp yeah can't recall the exact details, but I think the construct of allowing the filter to pass back the original reference or something held internally will be flexible and can be high perf if needed.

Reply:
Now I'm almost persuaded the original API I was commenting on was simpler than the alternatives here; my main goal is conceptual simplicity so that someone who is not familiar with the problem being solve can easily grok what these extensions points are. I'm persuaded by @snowp reply that both pre/post make sense for the intended application. The only suggestion I have on the API front there is to clearly explain the different scenarios you might use one vs. the other.